In [37]:
# Dependencies and Setup
import pandas as pd
from pathlib import Path

# File to Load (Remember to Change These)
school_data_to_load = Path("Resources/schools_complete.csv")
student_data_to_load = Path("Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head(10)

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   
5           5      Bryan Miranda      M   9th  Huang High School   
6           6      Sheena Carter      F  11th  Huang High School   
7           7       Nicole Baker      F  12th  Huang High School   
8           8       Michael Roth      M  10th  Huang High School   
9           9     Matthew Greene      M  10th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635  
5             94          94          0  District  2917  1910635  
6             82          80          0  District  2917  1910635  
7             96          69          0  District  2917  1910635  
8             95          87          0  District  2917  1910635  
9             96          84          0  District  2917  1910635

In [38]:
# Calculate the total number of unique schools
school_count = len(school_data_complete['school_name'].unique())
school_count

15

In [39]:
# Calculate the total number of students
student_count = school_data_complete['Student ID'].count()
student_count

39170

In [40]:
# Calculate the total budget
#total_budget = 
#total_budget
total_budget = school_data['budget'].sum()
total_budget

24649428

In [41]:
# Calculate the average (mean) math score
average_math_score = student_data['math_score'].mean()
average_math_score

78.98537145774827

In [42]:
# Calculate the average (mean) reading score
average_reading_score = student_data['reading_score'].mean()
average_reading_score


81.87784018381414

In [43]:
# Use the following to calculate the percentage of students who passed math (math scores greather than or equal to 70)
passing_math_count = school_data_complete[(school_data_complete["math_score"] >= 70)].count()["student_name"]
student_count = school_data_complete['Student ID'].unique()
#print(passing_math_count)
#print(len(student_count))

passing_math_percentage = passing_math_count / len(student_count) * 100
passing_math_percentage


74.9808526933878

In [44]:
# Calculate the percentage of students who passeed reading (hint: look at how the math percentage was calculated)  
passing_reading_count = school_data_complete[(school_data_complete["reading_score"] >= 70)].count()["student_name"]
student_count = school_data_complete['Student ID'].unique()
#print(passing_reading_count)
#print(len(student_count))

passing_reading_percentage = passing_reading_count / len(student_count) * 100
passing_reading_percentage

85.80546336482001

In [45]:
# Use the following to calculate the percentage of students that passed math and reading
passing_math_reading_count = school_data_complete[
    (school_data_complete["math_score"] >= 70) & (school_data_complete["reading_score"] >= 70)
].count()["student_name"]
student_count = school_data_complete['Student ID'].unique()
overall_passing_rate = passing_math_reading_count /  len(student_count) * 100
overall_passing_rate

65.17232575950983

In [46]:
# Create a high-level snapshot of the district's key metrics in a DataFrame
district_summary = pd.DataFrame({
    'Total Schools': [school_count],
    'Total Students': [student_count],
    'Total Budget':[total_budget],
    'Average Math Score':[average_math_score],
    'Average Reading Score':[average_reading_score],
    '% Passing Math':[passing_math_percentage],
    '% Passing Reading':[passing_reading_percentage],
    '% Overall Passing': [overall_passing_rate]
})
district_summary

Total Schools                                     Total Students  \
0             15  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...   

   Total Budget  Average Math Score  Average Reading Score  % Passing Math  \
0      24649428           78.985371               81.87784       74.980853   

   % Passing Reading  % Overall Passing  
0          85.805463          65.172326

In [47]:
#school name
school_name = (school_data_complete['school_name'].unique())
school_name

array(['Huang High School', 'Figueroa High School', 'Shelton High School',
       'Hernandez High School', 'Griffin High School',
       'Wilson High School', 'Cabrera High School', 'Bailey High School',
       'Holden High School', 'Pena High School', 'Wright High School',
       'Rodriguez High School', 'Johnson High School', 'Ford High School',
       'Thomas High School'], dtype=object)

In [48]:
#school type
school_types = school_data.set_index(["school_name"])["type"]
school_types


school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

In [49]:
#total students,
school_size_students = school_data.set_index(["school_name"])["size"]
school_size_students


school_name
Huang High School        2917
Figueroa High School     2949
Shelton High School      1761
Hernandez High School    4635
Griffin High School      1468
Wilson High School       2283
Cabrera High School      1858
Bailey High School       4976
Holden High School        427
Pena High School          962
Wright High School       1800
Rodriguez High School    3999
Johnson High School      4761
Ford High School         2739
Thomas High School       1635
Name: size, dtype: int64

In [50]:
#school budget
per_school_budget = school_data.set_index(["school_name"])["budget"]
per_school_budget

school_name
Huang High School        1910635
Figueroa High School     1884411
Shelton High School      1056600
Hernandez High School    3022020
Griffin High School       917500
Wilson High School       1319574
Cabrera High School      1081356
Bailey High School       3124928
Holden High School        248087
Pena High School          585858
Wright High School       1049400
Rodriguez High School    2547363
Johnson High School      3094650
Ford High School         1763916
Thomas High School       1043130
Name: budget, dtype: int64

In [51]:
#per student budget
per_student_budget = school_data.set_index(["school_name"])["budget"] / school_data.set_index(["school_name"])["size"]
per_student_budget

school_name
Huang High School        655.0
Figueroa High School     639.0
Shelton High School      600.0
Hernandez High School    652.0
Griffin High School      625.0
Wilson High School       578.0
Cabrera High School      582.0
Bailey High School       628.0
Holden High School       581.0
Pena High School         609.0
Wright High School       583.0
Rodriguez High School    637.0
Johnson High School      650.0
Ford High School         644.0
Thomas High School       638.0
dtype: float64

In [53]:

reduced = school_data_complete[['school_name','Student ID','math_score']]
testing = reduced.groupby('school_name').agg({'Student ID':'count', 'math_score': lambda x: (x>=70).sum()})

math_score = (testing['math_score'] / testing['Student ID']) * 100
passing_math_score = math_score
passing_math_score

school_name
Bailey High School       66.680064
Cabrera High School      94.133477
Figueroa High School     65.988471
Ford High School         68.309602
Griffin High School      93.392371
Hernandez High School    66.752967
Holden High School       92.505855
Huang High School        65.683922
Johnson High School      66.057551
Pena High School         94.594595
Rodriguez High School    66.366592
Shelton High School      93.867121
Thomas High School       93.272171
Wilson High School       93.867718
Wright High School       93.333333
dtype: float64

In [54]:
#% passing reading (the percentage of students who passed reading)
reduced = school_data_complete[['school_name','Student ID','reading_score']]
testing = reduced.groupby('school_name').agg({'Student ID':'count', 'reading_score': lambda x: (x>=70).sum()})

reading_score = (testing['reading_score'] / testing['Student ID']) * 100
passing_reading_score = reading_score
passing_reading_score

school_name
Bailey High School       81.933280
Cabrera High School      97.039828
Figueroa High School     80.739234
Ford High School         79.299014
Griffin High School      97.138965
Hernandez High School    80.862999
Holden High School       96.252927
Huang High School        81.316421
Johnson High School      81.222432
Pena High School         95.945946
Rodriguez High School    80.220055
Shelton High School      95.854628
Thomas High School       97.308869
Wilson High School       96.539641
Wright High School       96.611111
dtype: float64

In [55]:
#Average math score 
reduced_data = school_data_complete[['school_name','math_score']]
average_math_score = reduced_data.groupby('school_name').mean().squeeze()
average_math_score


school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64

In [56]:
#Average reading score
reduced_data = school_data_complete[['school_name','reading_score']]
average_reading_score = reduced_data.groupby('school_name').mean().squeeze()
average_reading_score

school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.725724
Thomas High School       83.848930
Wilson High School       83.989488
Wright High School       83.955000
Name: reading_score, dtype: float64

In [57]:
# %overall passing (the percentage of students who passed math AND reading)--student cnt that passed math & read with >=70 / school student cnt
#math_reading = school_data_complete_df[['Student ID','school_name','reading_score','math_score']]
#math_reading
passing_df = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]
passing_df
grouped_df = passing_df.groupby("school_name").agg({'Student ID': 'count', 'math_score': 'count', 'reading_score': 'count'})
grouped_df
merged_df = pd.merge(grouped_df, school_data_complete_df.groupby("school_name").agg({'Student ID': 'count'}), on="school_name")
merged_df
merged_df["mean_passing_score"] = (merged_df["math_score"] + merged_df["reading_score"]) / 2
merged_df["mean_passing_score"]
merged_df["passing_rate_per_student"] = (merged_df["mean_passing_score"] / merged_df["Student ID_y"])*100
overall_passing = merged_df["passing_rate_per_student"]
overall_passing


school_name
Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
Name: passing_rate_per_student, dtype: float64

In [58]:
top_schools = merged_df = merged_df.sort_values('passing_rate_per_student', ascending=False)
top_schools.head()



Student ID_x  math_score  reading_score  Student ID_y  \
school_name                                                                  
Cabrera High School          1697        1697           1697          1858   
Thomas High School           1487        1487           1487          1635   
Griffin High School          1330        1330           1330          1468   
Wilson High School           2068        2068           2068          2283   
Pena High School              871         871            871           962   

                     mean_passing_score  passing_rate_per_student  
school_name                                                        
Cabrera High School              1697.0                 91.334769  
Thomas High School               1487.0                 90.948012  
Griffin High School              1330.0                 90.599455  
Wilson High School               2068.0                 90.582567  
Pena High School                  871.0                 90.540541

In [59]:
bottom_schools = merged_df = merged_df.sort_values('passing_rate_per_student', ascending=True)
bottom_schools.head()


Student ID_x  math_score  reading_score  Student ID_y  \
school_name                                                                    
Rodriguez High School          2119        2119           2119          3999   
Figueroa High School           1569        1569           1569          2949   
Huang High School              1561        1561           1561          2917   
Hernandez High School          2481        2481           2481          4635   
Johnson High School            2549        2549           2549          4761   

                       mean_passing_score  passing_rate_per_student  
school_name                                                          
Rodriguez High School              2119.0                 52.988247  
Figueroa High School               1569.0                 53.204476  
Huang High School                  1561.0                 53.513884  
Hernandez High School              2481.0                 53.527508  
Johnson High School                2549.0                 53.539172

In [23]:
# Create a DataFrame called `per_school_summary` with columns for the calculations above.
summary_df = pd.DataFrame({
            #"School Name": school_name,
            "School Type": school_types,
            "TTL Students": school_size_students,
            "TTL School Budget":per_school_budget,
            "Per Student Budget": per_student_budget,
            "AVG Math Score":average_math_score,
            "AVG Reading Score":average_reading_score,
            "% Passing Math":passing_math_score,
            "% Passing Reading":passing_reading_score,
            "% Overall Passing":overall_passing,
            })

summary_df


School Type  TTL Students  TTL School Budget  \
school_name                                                          
Bailey High School       District          4976            3124928   
Cabrera High School       Charter          1858            1081356   
Figueroa High School     District          2949            1884411   
Ford High School         District          2739            1763916   
Griffin High School       Charter          1468             917500   
Hernandez High School    District          4635            3022020   
Holden High School        Charter           427             248087   
Huang High School        District          2917            1910635   
Johnson High School      District          4761            3094650   
Pena High School          Charter           962             585858   
Rodriguez High School    District          3999            2547363   
Shelton High School       Charter          1761            1056600   
Thomas High School        Charter          1635            1043130   
Wilson High School        Charter          2283            1319574   
Wright High School        Charter          1800            1049400   

                       Per Student Budget  AVG Math Score  AVG Reading Score  \
school_name                                                                    
Bailey High School                  628.0       77.048432          81.033963   
Cabrera High School                 582.0       83.061895          83.975780   
Figueroa High School                639.0       76.711767          81.158020   
Ford High School                    644.0       77.102592          80.746258   
Griffin High School                 625.0       83.351499          83.816757   
Hernandez High School               652.0       77.289752          80.934412   
Holden High School                  581.0       83.803279          83.814988   
Huang High School                   655.0       76.629414          81.182722   
Johnson High School                 650.0       77.072464          80.966394   
Pena High School                    609.0       83.839917          84.044699   
Rodriguez High School               637.0       76.842711          80.744686   
Shelton High School                 600.0       83.359455          83.725724   
Thomas High School                  638.0       83.418349          83.848930   
Wilson High School                  578.0       83.274201          83.989488   
Wright High School                  583.0       83.682222          83.955000   

                       % Passing Math  % Passing Reading  % Overall Passing  
school_name                                                                  
Bailey High School          66.680064          81.933280          54.642283  
Cabrera High School         94.133477          97.039828          91.334769  
Figueroa High School        65.988471          80.739234          53.204476  
Ford High School            68.309602          79.299014          54.289887  
Griffin High School         93.392371          97.138965          90.599455  
Hernandez High School       66.752967          80.862999          53.527508  
Holden High School          92.505855          96.252927          89.227166  
Huang High School           65.683922          81.316421          53.513884  
Johnson High School         66.057551          81.222432          53.539172  
Pena High School            94.594595          95.945946          90.540541  
Rodriguez High School       66.366592          80.220055          52.988247  
Shelton High School         93.867121          95.854628          89.892107  
Thomas High School          93.272171          97.308869          90.948012  
Wilson High School          93.867718          96.539641          90.582567  
Wright High School          93.333333          96.611111          90.333333

In [64]:
# this works in the pycitychools_starter
math_grade_df = school_data_complete.groupby(['school_name', 'grade']).mean()
math_grade_avg_df = math_grade_df.pivot_table(index='school_name', columns='grade', values='math_score')
math_grade_avg_order_df = math_grade_avg_df.reindex(sorted(math_grade_avg_df.columns,reverse=True),axis=1)
math_grade_avg_order_df

grade                        9th       12th       11th       10th
school_name                                                      
Bailey High School     77.083676  76.492218  77.515588  76.996772
Cabrera High School    83.094697  83.277487  82.765560  83.154506
Figueroa High School   76.403037  77.151369  76.884344  76.539974
Ford High School       77.361345  76.179963  76.918058  77.672316
Griffin High School    82.044010  83.356164  83.842105  84.229064
Hernandez High School  77.438495  77.186567  77.136029  77.337408
Holden High School     83.787402  82.855422  85.000000  83.429825
Huang High School      77.027251  77.225641  76.446602  75.908735
Johnson High School    77.187857  76.863248  77.491653  76.691117
Pena High School       83.625455  84.121547  84.328125  83.372000
Rodriguez High School  76.859966  77.690748  76.395626  76.612500
Shelton High School    83.420755  83.778976  83.383495  82.917411
Thomas High School     83.590022  83.497041  83.498795  83.087886
Wilson High School     83.085578  83.035794  83.195326  83.724422
Wright High School     83.264706  83.644986  83.836782  84.010288

In [65]:
# this works in the pycitychools_starter
reading_grade_df = school_data_complete.groupby(['school_name', 'grade']).mean()
reading_grade_avg_df = reading_grade_df.pivot_table(index='school_name', columns='grade', values='reading_score')
reading_grade_avg_order_df = reading_grade_avg_df.reindex(sorted(reading_grade_avg_df.columns,reverse=True),axis=1)
reading_grade_avg_order_df

grade                        9th       12th       11th       10th
school_name                                                      
Bailey High School     81.303155  80.912451  80.945643  80.907183
Cabrera High School    83.676136  84.287958  83.788382  84.253219
Figueroa High School   81.198598  81.384863  80.640339  81.408912
Ford High School       80.632653  80.662338  80.403642  81.262712
Griffin High School    83.369193  84.013699  84.288089  83.706897
Hernandez High School  80.866860  80.857143  81.396140  80.660147
Holden High School     83.677165  84.698795  83.815534  83.324561
Huang High School      81.290284  80.305983  81.417476  81.512386
Johnson High School    81.260714  81.227564  80.616027  80.773431
Pena High School       83.807273  84.591160  84.335938  83.612000
Rodriguez High School  80.993127  80.376426  80.864811  80.629808
Shelton High School    84.122642  82.781671  84.373786  83.441964
Thomas High School     83.728850  83.831361  83.585542  84.254157
Wilson High School     83.939778  84.317673  83.764608  84.021452
Wright High School     83.833333  84.073171  84.156322  83.812757

In [70]:
spending_bins = [0, 585, 630, 645, 680]
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

summary_df['PSS Reclass'] = pd.cut(summary_df['Per Student Budget'], bins=spending_bins, labels=labels)
summary_reclass_df = summary_df
summary_reclass_df

School Type  TTL Students  TTL School Budget  \
school_name                                                          
Bailey High School       District          4976            3124928   
Cabrera High School       Charter          1858            1081356   
Figueroa High School     District          2949            1884411   
Ford High School         District          2739            1763916   
Griffin High School       Charter          1468             917500   
Hernandez High School    District          4635            3022020   
Holden High School        Charter           427             248087   
Huang High School        District          2917            1910635   
Johnson High School      District          4761            3094650   
Pena High School          Charter           962             585858   
Rodriguez High School    District          3999            2547363   
Shelton High School       Charter          1761            1056600   
Thomas High School        Charter          1635            1043130   
Wilson High School        Charter          2283            1319574   
Wright High School        Charter          1800            1049400   

                       Per Student Budget  AVG Math Score  AVG Reading Score  \
school_name                                                                    
Bailey High School                  628.0       77.048432          81.033963   
Cabrera High School                 582.0       83.061895          83.975780   
Figueroa High School                639.0       76.711767          81.158020   
Ford High School                    644.0       77.102592          80.746258   
Griffin High School                 625.0       83.351499          83.816757   
Hernandez High School               652.0       77.289752          80.934412   
Holden High School                  581.0       83.803279          83.814988   
Huang High School                   655.0       76.629414          81.182722   
Johnson High School                 650.0       77.072464          80.966394   
Pena High School                    609.0       83.839917          84.044699   
Rodriguez High School               637.0       76.842711          80.744686   
Shelton High School                 600.0       83.359455          83.725724   
Thomas High School                  638.0       83.418349          83.848930   
Wilson High School                  578.0       83.274201          83.989488   
Wright High School                  583.0       83.682222          83.955000   

                       % Passing Math  % Passing Reading  % Overall Passing  \
school_name                                                                   
Bailey High School          66.680064          81.933280          54.642283   
Cabrera High School         94.133477          97.039828          91.334769   
Figueroa High School        65.988471          80.739234          53.204476   
Ford High School            68.309602          79.299014          54.289887   
Griffin High School         93.392371          97.138965          90.599455   
Hernandez High School       66.752967          80.862999          53.527508   
Holden High School          92.505855          96.252927          89.227166   
Huang High School           65.683922          81.316421          53.513884   
Johnson High School         66.057551          81.222432          53.539172   
Pena High School            94.594595          95.945946          90.540541   
Rodriguez High School       66.366592          80.220055          52.988247   
Shelton High School         93.867121          95.854628          89.892107   
Thomas High School          93.272171          97.308869          90.948012   
Wilson High School          93.867718          96.539641          90.582567   
Wright High School          93.333333          96.611111          90.333333   

                      PSS Reclass  
school_name                        
Bailey High School       $585-630  
Cabrera High Schoo

In [89]:
spending_math_scores = summary_reclass_df.groupby(["school_name","PSS Reclass"]).mean()["AVG Math Score"]
spending_reading_scores = summary_reclass_df.groupby(["school_name","PSS Reclass"]).mean()["AVG Reading Score"]
spending_passing_math = summary_reclass_df.groupby(["school_name","PSS Reclass"]).mean()["% Passing Math"]
spending_passing_reading = summary_reclass_df.groupby(["school_name","PSS Reclass"]).mean()["% Passing Reading"]
overall_passing_spending = summary_reclass_df.groupby(["school_name","PSS Reclass"]).mean()["% Overall Passing"]

summary_spending_df = pd.DataFrame({
    'AVG Math Score': spending_math_scores,
    'AVG Reading Score': spending_reading_scores,
    '% Passing Math': spending_passing_math,
    '% Passing Reading': spending_passing_reading,
    '% Overall Passing': overall_passing_spending
})

summary_spending_df.reset_index(inplace=True)
summary_spending_df = summary_spending_df.dropna(subset=['% Overall Passing'])
summary_spending_df

school_name PSS Reclass  AVG Math Score  AVG Reading Score  \
1      Bailey High School    $585-630       77.048432          81.033963   
4     Cabrera High School       <$585       83.061895          83.975780   
10   Figueroa High School    $630-645       76.711767          81.158020   
14       Ford High School    $630-645       77.102592          80.746258   
17    Griffin High School    $585-630       83.351499          83.816757   
23  Hernandez High School    $645-680       77.289752          80.934412   
24     Holden High School       <$585       83.803279          83.814988   
31      Huang High School    $645-680       76.629414          81.182722   
35    Johnson High School    $645-680       77.072464          80.966394   
37       Pena High School    $585-630       83.839917          84.044699   
42  Rodriguez High School    $630-645       76.842711          80.744686   
45    Shelton High School    $585-630       83.359455          83.725724   
50     Thomas High School    $630-645       83.418349          83.848930   
52     Wilson High School       <$585       83.274201          83.989488   
56     Wright High School       <$585       83.682222          83.955000   

    % Passing Math  % Passing Reading  % Overall Passing  
1        66.680064          81.933280          54.642283  
4        94.133477          97.039828          91.334769  
10       65.988471          80.739234          53.204476  
14       68.309602          79.299014          54.289887  
17       93.392371          97.138965          90.599455  
23       66.752967          80.862999          53.527508  
24       92.505855          96.252927          89.227166  
31       65.683922          81.316421          53.513884  
35       66.057551          81.222432          53.539172  
37       94.594595          95.945946          90.540541  
42       66.366592          80.220055          52.988247  
45       93.867121          95.854628          89.892107  
50       93.272171          97.308869          90.948012  
52       93.867718          96.539641          90.582567  
56       93.333333          96.611111          90.333333

In [91]:
size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

summary_df['PSS Reclass'] = pd.cut(summary_df['TTL Students'], bins=size_bins, labels=labels)
summary_reclass_df = summary_df
summary_reclass_df


School Type  TTL Students  TTL School Budget  \
school_name                                                          
Bailey High School       District          4976            3124928   
Cabrera High School       Charter          1858            1081356   
Figueroa High School     District          2949            1884411   
Ford High School         District          2739            1763916   
Griffin High School       Charter          1468             917500   
Hernandez High School    District          4635            3022020   
Holden High School        Charter           427             248087   
Huang High School        District          2917            1910635   
Johnson High School      District          4761            3094650   
Pena High School          Charter           962             585858   
Rodriguez High School    District          3999            2547363   
Shelton High School       Charter          1761            1056600   
Thomas High School        Charter          1635            1043130   
Wilson High School        Charter          2283            1319574   
Wright High School        Charter          1800            1049400   

                       Per Student Budget  AVG Math Score  AVG Reading Score  \
school_name                                                                    
Bailey High School                  628.0       77.048432          81.033963   
Cabrera High School                 582.0       83.061895          83.975780   
Figueroa High School                639.0       76.711767          81.158020   
Ford High School                    644.0       77.102592          80.746258   
Griffin High School                 625.0       83.351499          83.816757   
Hernandez High School               652.0       77.289752          80.934412   
Holden High School                  581.0       83.803279          83.814988   
Huang High School                   655.0       76.629414          81.182722   
Johnson High School                 650.0       77.072464          80.966394   
Pena High School                    609.0       83.839917          84.044699   
Rodriguez High School               637.0       76.842711          80.744686   
Shelton High School                 600.0       83.359455          83.725724   
Thomas High School                  638.0       83.418349          83.848930   
Wilson High School                  578.0       83.274201          83.989488   
Wright High School                  583.0       83.682222          83.955000   

                       % Passing Math  % Passing Reading  % Overall Passing  \
school_name                                                                   
Bailey High School          66.680064          81.933280          54.642283   
Cabrera High School         94.133477          97.039828          91.334769   
Figueroa High School        65.988471          80.739234          53.204476   
Ford High School            68.309602          79.299014          54.289887   
Griffin High School         93.392371          97.138965          90.599455   
Hernandez High School       66.752967          80.862999          53.527508   
Holden High School          92.505855          96.252927          89.227166   
Huang High School           65.683922          81.316421          53.513884   
Johnson High School         66.057551          81.222432          53.539172   
Pena High School            94.594595          95.945946          90.540541   
Rodriguez High School       66.366592          80.220055          52.988247   
Shelton High School         93.867121          95.854628          89.892107   
Thomas High School          93.272171          97.308869          90.948012   
Wilson High School          93.867718          96.539641          90.582567   
Wright High School          93.333333          96.611111          90.333333   

                              PSS Reclass  
school_name                                
Bailey High School      Large (2000-50

In [92]:
spending_math_scores = summary_reclass_df.groupby(["school_name","TTL Students"]).mean()["AVG Math Score"]
spending_reading_scores = summary_reclass_df.groupby(["school_name","TTL Students"]).mean()["AVG Reading Score"]
spending_passing_math = summary_reclass_df.groupby(["school_name","TTL Students"]).mean()["% Passing Math"]
spending_passing_reading = summary_reclass_df.groupby(["school_name","TTL Students"]).mean()["% Passing Reading"]
overall_passing_spending = summary_reclass_df.groupby(["school_name","TTL Students"]).mean()["% Overall Passing"]

summary_spending_df = pd.DataFrame({
    'AVG Math Score': spending_math_scores,
    'AVG Reading Score': spending_reading_scores,
    '% Passing Math': spending_passing_math,
    '% Passing Reading': spending_passing_reading,
    '% Overall Passing': overall_passing_spending
})

summary_spending_df.reset_index(inplace=True)
summary_spending_df = summary_spending_df.dropna(subset=['% Overall Passing'])
summary_spending_df

school_name  TTL Students  AVG Math Score  AVG Reading Score  \
0      Bailey High School          4976       77.048432          81.033963   
1     Cabrera High School          1858       83.061895          83.975780   
2    Figueroa High School          2949       76.711767          81.158020   
3        Ford High School          2739       77.102592          80.746258   
4     Griffin High School          1468       83.351499          83.816757   
5   Hernandez High School          4635       77.289752          80.934412   
6      Holden High School           427       83.803279          83.814988   
7       Huang High School          2917       76.629414          81.182722   
8     Johnson High School          4761       77.072464          80.966394   
9        Pena High School           962       83.839917          84.044699   
10  Rodriguez High School          3999       76.842711          80.744686   
11    Shelton High School          1761       83.359455          83.725724   
12     Thomas High School          1635       83.418349          83.848930   
13     Wilson High School          2283       83.274201          83.989488   
14     Wright High School          1800       83.682222          83.955000   

    % Passing Math  % Passing Reading  % Overall Passing  
0        66.680064          81.933280          54.642283  
1        94.133477          97.039828          91.334769  
2        65.988471          80.739234          53.204476  
3        68.309602          79.299014          54.289887  
4        93.392371          97.138965          90.599455  
5        66.752967          80.862999          53.527508  
6        92.505855          96.252927          89.227166  
7        65.683922          81.316421          53.513884  
8        66.057551          81.222432          53.539172  
9        94.594595          95.945946          90.540541  
10       66.366592          80.220055          52.988247  
11       93.867121          95.854628          89.892107  
12       93.272171          97.308869          90.948012  
13       93.867718          96.539641          90.582567  
14       93.333333          96.611111          90.333333

A total of 15 schools were analyzed for test results in the area of mathematics and reading for high school students grades from 9th to 12th. The analysis covers what the students averages scores were for math and reading and what were the passing scores and percents for students above 70 in the mentioned areas. Out of the 15 schools, 8 of them are Charter type while the remaining 7 are District type. In addition, budget amounts and school size (student counts) were also looked at to help point out any correlation in regards to the final analysis.

The student count for the Distric and Charter schools are 26,976 and 12,194 respectively. The budgets are $17,347,923 for the District while the Charter budget is $7,301,505. This in general states that Discrict schools have a per student budget of $643 vs Charter at $599, although in the analysis each school has its own budget per student for further details. With the results of this analysis an observation can be made in regards to if the "per student spend" equates to a better passing or scoring grades. The analysis also helps to compare if school size also has any impact in students overall performance.

Charter schools showed that 89% of the students had an overall passing in both subjects of 70% scoring while the Districts under performed to about 55%. What really interesting is that the Math scores for the District are very "tight" on an average of about 77% per school and the Charter showed the same "tightness" per school but with an 83.5%. Meaning that the Charter schools in general are having the students score about 7 points higher in math. In regards to reading, the same "tightness" can be seen in both school types, but with a variance of under 2.75 points and not 7 like in math. This means that the math program is superior in Charter than District schools and the reading program is much closer to being about the same.

Another interesting observation is that the average math scores in the District are higher than the passing over 70%. While the Charter is showing the opposite, the over 70% passing is higher than the average math score. In the reading aspect, the District average reading scores remain just about the same as the passing over 70%, while the Charter just like in the math area is showing a better over 70% results than the average score. 

So, without question, this analysis is showing that the Charter are yielding a better education than the District schools. Since there are far more students in the District schools compared to the Charter, its probable that there are more students in a single class for one teacher than in the Charter school, so it would be interesting to determine student per teacher ratio. Lastly, in regards to the budget...having a budget per student of $643 with a passing of 55% in the District vs $599 with a passing of 89% should emphasize that spending more does not equate a "better" education. So, the District should look on how is the budget being spend and make the necessary changes to allocate in such a way that the student is benefiting from it thru an enhacement of grades.



